### ISSUE TITLE
Cell radius prop is wrongly typed

### ISSUE THREAD

**Issue by CyrilQuandalle** [recharts/recharts#3325]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[http://noLinkNeeded](http://noLinkNeeded)

### Steps to reproduce
Try to use the component Cell and pass an array of numbers for example : [8, 8, 0,0] to the radius prop 

### What is expected?
The prop radius of the Cell should be typed to accept an array of numbers since the behaviour of the component seems to be ok while passing one.

### What is actually happening?
I have a Typescript error saying that it should only be a number instead of an array of number

| Environment | Info |
|---|---|
| Recharts | v2.3.2 |
| React | 17 |
| System | Ubuntu 22.4 |
| Browser | chrome 109 |

**Comment by ckifer** [comment]
https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/r#

Radius is supposed to be a length (number with or without units) or percentage, not sure how it is working with an array but it shouldn't

**Comment by CyrilQuandalle** [comment]
Hello,
Thank you for your quick answer.
The thing that I want to do is to the set the top two border-radius of a `Bar` in a `Barchart` conditionally depending if there is some data in the other bars of the chart. 
If I set the radius with one number like you said it sets the radius on all four corners of my bars. 
I think in my case its working because my `Cell` component is used as a child of a `Bar` component.
`<Bar
       dataKey={'someDataKey'}
              stackId="a"
              fill="url(#colorData)"
            >
              {data.map((entry, i) => {
                return (
                  <Cell
                    key={i}
                   // here I have a TS error
                    radius={
                      ( someCondition ? [8, 8, 0, 0] : [0, 0, 0, 0]
                    }
                  />
                );
              })}
            </Bar>`

**Comment by ckifer** [comment]
Thank you, the context helps. (please consider adding a codesandbox example next time you submit an issue so its easier to look at up front - thanks!)

Repro:
https://codesandbox.io/s/bar-chart-with-radius-type-err-uo7inb

So looks like it works on `Bar` itself - do you need to conditionally change the radius in the same series?

**Comment by CyrilQuandalle** [comment]
Thanks for the codesandbox repro. (I tried to make one but wasn't able to have the TS error ;-))
My use case is still a bit different, I'm doing a stacked barchart so whenever there is no data in other bars I need to have to have some rounded radius on the last one with data.
Here's what it looks like : 
![image](https://user-images.githubusercontent.com/8200968/216667746-7e8c2c2c-3050-4381-b07e-852d0cbc38de.png)
So sometimes its the green that is rounded when there is no red and sometimes its the blue when the other two don't have data.

**Comment by ckifer** [comment]
I see, thanks for the screenshot - modified the codesandbox to reflect a stacked bar.

Tbh I'm not sure what this entails as `radius` doesn't apply to everything that uses `Cell` and `Cell` isn't generic at the moment. I will label this bug / enhancement but can't promise it will be worked on any time soon. Feel free to pull down the code take a stab at it if you want, but I wouldn't want to accept an array for elements where an array will have no effect/cause the same time of problem

You could try rendering `Rectangle` element's instead of `Cell` - those do have the `radius` prop (but then you would need the position values).

**Comment by CyrilQuandalle** [comment]
Thank you, it's ok if it's not fixed right away since it's not really blocking me.
I'll take a look at the `Rectangle` and  `Customized` even if the `Cell` seems more attractive since it works out of box.
Thank you very much for your help and your time  ;-).

**Comment by jeffnmock** [comment]
> Thank you, it's ok if it's not fixed right away since it's not really blocking me. I'll take a look at the `Rectangle` and `Customized` even if the `Cell` seems more attractive since it works out of box. Thank you very much for your help and your time ;-).

Don't suppose you had any luck getting this to work? I'm also trying to figure out how to add a radius on only the last segment of a stacked bar.

**Comment by CyrilQuandalle** [comment]
Hello @jeffnmock ,
My solution mentioned there works with `//@ts-ignore` before of course. You just have to replace `someCondition` by your own

> Hello, Thank you for your quick answer. The thing that I want to do is to the set the top two border-radius of a `Bar` in a `Barchart` conditionally depending if there is some data in the other bars of the chart. If I set the radius with one number like you said it sets the radius on all four corners of my bars. I think in my case its working because my `Cell` component is used as a child of a `Bar` component. `<Bar dataKey={'someDataKey'} stackId="a" fill="url(#colorData)" > {data.map((entry, i) => { return ( <Cell key={i} // here I have a TS error radius={ ( someCondition ? [8, 8, 0, 0] : [0, 0, 0, 0] } /> ); })} </Bar>`

**Comment by jeffnmock** [comment]
Hi @CyrilQuandalle, I see there's no built in way. We just have to manually calculate whether each bar should be rounded based on the rest of the data. Thanks!

**Comment by CyrilQuandalle** [comment]
Yes you just have to use what you retrieve from each `entry` element while doing  `data.map`

**Comment by marcinwadon** [comment]
I support [this](https://github.com/recharts/recharts/issues/3325#issuecomment-1416006713) comment.
Since the `Cell` `radius` property is from `SVGAttributes`, there should be no array assigned to it.
If you want to dynamically select Cells to make a radius (e.g. top bottom part) you can do the opposite: assign a radius to each and every `Bar` component and set `radius={0}` to `Cell` conditionally, by negating the existing condition. `Cell` `radius` property has more importance hence would overwrite the `Bar` radius.

[https://codesandbox.io/s/bar-chart-with-radius-type-err-forked-d7yx3r?file=/src/App.tsx](https://codesandbox.io/s/bar-chart-with-radius-type-err-forked-d7yx3r?file=/src/App.tsx)

**Comment by nikolasrieble** [comment]
@marcinwadon Great solution, thank you for sharing. So really, this issue should be closed then? 
wdyt @ckifer ?

**Comment by ckifer** [comment]
I think either this should be more obvious from the docs or we should make it easier to do correctly. I.e if we intercepted radius, made it acceptable to be an array, and did this for the user

**Comment by hienngm** [comment]
This is my workaround. Hope it helps.
```
  <Bar
      key={channel.name}
      dataKey={channel.name}
      fill={getStrokeColor({ index, channel: channel.name })}
      stackId="viewCount"
      radius={[6, 6, 0, 0]}
    >
      {chartData.map((entry, index) => {
        const keys = Object.keys(entry);
        const values = Object.values(entry);

        const channelNameIndex = keys.findIndex((key) => key === channel.name);
        const lastBarIndex = values.findLastIndex((value) => value !== 0);

        if (channelNameIndex === lastBarIndex) {
          return <Cell key={`cell-${index}`} />;
        }

        return <Cell key={`cell-${index}`} radius={0} />;
      })}
    </Bar>
```
<img width="1262" alt="image" src="https://github.com/recharts/recharts/assets/112270783/02c9ac19-e740-42d3-ac2e-54d519c41852">

**Comment by hienngm** [comment]
This is my workaround, hope it helps https://github.com/recharts/recharts/issues/3887#issuecomment-2208480245

**Comment by luap2703** [comment]
Imo it would be most convenient if it does accept the array as input. 

Radius is natively also not a prop on the svg element (so its custom behaviour in any way, right?) and I would assume in 99.9% of cases when ppl using the radius prop on the Cell element in recharts, it's about manually adjusting the radius on each bar.

For our use-case (dynamically setting top AND/OR bottom, depending whether the bar is the topmost AND/OR bottom-most bar => 4 possible scenarios) the proposed solution by @marcinwadon doesn't work, as it allows only binary states (either set the radius or ultimately not).

**Comment by ckifer** [comment]
It looks like this is fixed in 3.x
https://github.com/recharts/recharts/blame/5ae977d0eaf7de06df56aa99ac95781d316fb850/src/cartesian/Bar.tsx#L112

This always supported an array in Rectangle
https://github.com/recharts/recharts/blame/5ae977d0eaf7de06df56aa99ac95781d316fb850/src/cartesian/Bar.tsx#L112

But it was overridden by `SVGAttributes` `radius` property.

If someone wants to backport the change in the first link to the master branch I can release it in the next patch, otherwise it'll be fixed in 3.0

**Comment by ckifer** [comment]
Fixed in 2.13.1, closing this

**Comment by IPJT** [comment]
To anyone else seeing this thread in the future. Doing what the initial author suggested in the first post is totally fine. 
I.e:
**Passing an array of numbers for example : [8, 8, 0,0] to the radius prop of the Cell-component is totally fine.** 
This works as the underlying Rectangle-component supports it (again, it's **not** a supported SVG-attribute).

I'm not really following why the typing-issue is fixed in [#5116](https://github.com/recharts/recharts/pull/5116) as suggested by @ckifer  as it modifies the Props of the Bar and not the Cell-component. I'm on 2.15.3 and still get the TS-error (which I'm ignoring)

**Comment by ckifer** [comment]
Yep, supported in Rectangle - forgive my lack of familiarity with the code 3 years ago. 

As for Cell it's more tricky to type Cell because there is no Cell specific to each chart type. Cell prop types for Bar are the same as Cell props for Scatter, etc. which don't have support for a radius prop.
